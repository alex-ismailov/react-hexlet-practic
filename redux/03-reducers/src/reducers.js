import _ from 'lodash';
import { combineReducers } from 'redux';

const comments = (state = {}, action) => {
  // BEGIN (write your solution here)
  switch (action.type) {
    case 'TASK_COMMENT_ADD': {
      const { id, taskId } = action.payload.comment;
      return { ...state, [id]: { id, taskId } };
    }
    case 'TASK_COMMENT_REMOVE': {
      const { id } = action.payload;
      return _.omit(state, id);
    }
    case 'TASK_REMOVE': {
      const { id } = action.payload;
      const idKeys = Object.keys(state)
        .filter((key) => state[key].taskId === id);

      return _.omit(state, idKeys);
    }
    default:
      return state;
  }
  // END
};

const tasks = (state = {}, action) => {
  // BEGIN (write your solution here)
  switch (action.type) {
    case 'TASK_ADD': {
      const { id } = action.payload.task;
      return { ...state, [id]: { id } };
    }
    case 'TASK_REMOVE': {
      const { id } = action.payload;
      return _.omit(state, id);
    }
    default:
      return state;
  }
  // END
};

export default combineReducers({
  comments,
  tasks,
});
