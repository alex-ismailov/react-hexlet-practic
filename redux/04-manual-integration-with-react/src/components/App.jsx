/* eslint-disable react/static-property-placement */
import React from 'react';

// BEGIN (write your solution here)
export default class App extends React.Component {
  static defaultProps = {
    text: '',
  };

  handleInputText = (e) => {
    e.preventDefault();
    const { dispatch, updateText } = this.props;
    const { target: { value } } = e;
    dispatch(updateText(value));
  };

  handleResetText = (e) => {
    e.preventDefault();
    const { dispatch, resetText } = this.props;
    dispatch(resetText());
  };

  render() {
    const { text } = this.props;
    
    return (
      <div>
        <form>
          <input onChange={this.handleInputText} type="text" value={text} />
          <button onClick={this.handleResetText} type="button">Reset</button>
        </form>
        {text && <div>{text}</div>}
      </div>
    );
  }
}
// END
