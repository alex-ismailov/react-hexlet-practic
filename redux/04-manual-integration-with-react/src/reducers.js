import { combineReducers } from 'redux';

const text = (state = '', action) => {
  // BEGIN (write your solution here)
  switch (action.type) {
    case 'TEXT_UPDATE': {
      const { text } = action.payload;
      return text;
    }
    case 'TEXT_RESET': {
      return '';
    }
    default:
      return state;
  }
  // END
};

export default combineReducers({
  text,
});
