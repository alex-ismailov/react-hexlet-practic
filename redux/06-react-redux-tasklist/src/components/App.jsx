import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import { updateNewTaskText, addTask, removeTask } from '../actions/index';

// BEGIN (write your solution here)
// const mapStateToProps = (state) => {
//   const props = {
//     text: state.text,
//     tasks: state.tasks,
//   };
//   return props;
// };

const mapStateToProps = ({ text, tasks }) => {
  const props = { text, tasks };
  return props;
};

class App extends React.Component {
  handleInputTaskText = (e) => {
    const { dispatch } = this.props;
    const { target: { value } } = e;
    dispatch(updateNewTaskText(value));
  };

  handleAddTask = (e) => {
    e.preventDefault();
    const { dispatch, text } = this.props;
    const newTask = { id: _.uniqueId(), text };
    dispatch(addTask(newTask));
  };

  handleRemoveTask = (id) => () => {
    const { dispatch } = this.props;
    dispatch(removeTask(id));
  };

  renderTasks() {
    const { tasks } = this.props;
    if (tasks.length === 0) {
      return null;
    }

    const taskItems = tasks.map((task) => (
      <li key={task.id} className="list-group-item d-flex">
        <span className="mr-auto">{task.text}</span>
        <button onClick={this.handleRemoveTask(task.id)} type="button" className="close">
          <span>&times;</span>
        </button>
      </li>
    ));

    return (
      <div className="mt-3">
        <ul className="list-group">{taskItems}</ul>
      </div>
    );
  }

  render() {
    const { text } = this.props;

    return (
      <div className="col-5">
        <form onSubmit={this.handleAddTask} action="" className="form-inline">
          <div className="form-group mx-sm-3">
            <input onChange={this.handleInputTaskText} type="text" required value={text} />
          </div>
          <button type="submit" className="btn btn-primary btn-sm">Add</button>
        </form>
        {this.renderTasks()}
      </div>
    );
  }
}

export default connect(mapStateToProps)(App);
// END
