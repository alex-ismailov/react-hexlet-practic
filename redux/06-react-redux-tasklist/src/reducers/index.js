
import { combineReducers } from 'redux';

const text = (state = '', action) => {
  switch (action.type) {
    case 'TEXT_UPDATE': {
      return action.payload.text;
    }
    case 'TASK_ADD': {
      return '';
    }
    default:
      return state;
  }
};

// BEGIN (write your solution here)
const tasks = (state = [], action) => {
  switch (action.type) {
    case 'TASK_ADD': {

      return;
    }
    case 'TASK_REMOVE': {

      return;
    }
    default:
      return state;
  }
};
// END

export default combineReducers({
  text,
  tasks,
});
