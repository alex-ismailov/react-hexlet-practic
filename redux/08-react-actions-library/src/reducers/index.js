import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from '../actions/index.js';

// BEGIN (write your solution here)

// END

export default combineReducers({
  taskText,
  tasks,
});
