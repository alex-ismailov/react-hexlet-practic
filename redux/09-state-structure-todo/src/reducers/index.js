import _ from 'lodash';
import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from '../actions/index.js';

const tasks = handleActions({
  [actions.addTask](state, { payload: { task } }) {
    const { byId, allIds } = state;
    return {
      byId: { ...byId, [task.id]: task },
      allIds: [task.id, ...allIds],
    };
  },
  [actions.removeTask](state, { payload: { id } }) {
    const { byId, allIds } = state;
    return {
      byId: _.omit(byId, id),
      allIds: _.without(allIds, id),
    };
  },
  // BEGIN (write your solution here)
  [actions.toggleTaskState]: (state, { payload: { id } }) => {
    const newTaskState = state.byId[id].state === 'active'
      ? 'finished'
      : 'active';

    const newState = { ...state };
    _.set(newState, ['byId', id, 'state'], newTaskState);

    return newState;
  },
  // END
}, { byId: {}, allIds: [] });

const text = handleActions({
  [actions.addTask]() {
    return '';
  },
  [actions.updateNewTaskText](_state, { payload }) {
    return payload.text;
  },
}, '');

export default combineReducers({
  tasks,
  text,
});
