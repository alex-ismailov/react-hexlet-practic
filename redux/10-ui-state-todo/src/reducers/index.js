import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from '../actions/index.js';

const tasks = handleActions({
  [actions.addTask](state, { payload: { task } }) {
    const { byId, allIds } = state;
    return {
      byId: { ...byId, [task.id]: task },
      allIds: [task.id, ...allIds],
    };
  },
}, { byId: {}, allIds: [] });

const tasksUIState = handleActions({
  // BEGIN (write your solution here)
  
  // END
}, {});

const text = handleActions({
  [actions.addTask]() {
    return '';
  },
  [actions.updateNewTaskText](state, { payload }) {
    return payload.text;
  },
}, '');

export default combineReducers({
  tasks,
  tasksUIState,
  text,
});
