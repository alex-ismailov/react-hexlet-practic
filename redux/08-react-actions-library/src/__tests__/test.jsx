import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import reducers from '../reducers/index.js';
import App from '../components/App.jsx';

test('Store', () => {
  const store = createStore(reducers);

  const vdom = (
    <Provider store={store}>
      <App />
    </Provider>
  );
  const { asFragment } = render(vdom);
  expect(asFragment()).toMatchSnapshot();

  const newTaskInput = screen.getByTestId('input');
  const newTaskSubmit = screen.getByTestId('submit');

  userEvent.type(newTaskInput, 'na-na');
  expect(asFragment()).toMatchSnapshot();

  userEvent.click(newTaskSubmit);
  expect(asFragment()).toMatchSnapshot();

  userEvent.type(newTaskInput, 'another task');
  expect(asFragment()).toMatchSnapshot();

  userEvent.click(newTaskSubmit);
  expect(asFragment()).toMatchSnapshot();

  const closeButtons = screen.getAllByText('Ã—');
  userEvent.click(closeButtons[0]);
  expect(asFragment()).toMatchSnapshot();
});
