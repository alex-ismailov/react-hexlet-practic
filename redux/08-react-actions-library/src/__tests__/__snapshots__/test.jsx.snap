// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store 1`] = `
<DocumentFragment>
  <div
    class="col-5"
  >
    <form
      action=""
      class="form-inline"
    >
      <div
        class="form-group mx-sm-3"
      >
        <input
          data-testid="input"
          required=""
          type="text"
          value=""
        />
      </div>
      <input
        class="btn btn-primary btn-sm"
        data-testid="submit"
        type="submit"
        value="Add"
      />
    </form>
  </div>
</DocumentFragment>
`;

exports[`Store 2`] = `
<DocumentFragment>
  <div
    class="col-5"
  >
    <form
      action=""
      class="form-inline"
    >
      <div
        class="form-group mx-sm-3"
      >
        <input
          data-testid="input"
          required=""
          type="text"
          value="na-na"
        />
      </div>
      <input
        class="btn btn-primary btn-sm"
        data-testid="submit"
        type="submit"
        value="Add"
      />
    </form>
  </div>
</DocumentFragment>
`;

exports[`Store 3`] = `
<DocumentFragment>
  <div
    class="col-5"
  >
    <form
      action=""
      class="form-inline"
    >
      <div
        class="form-group mx-sm-3"
      >
        <input
          data-testid="input"
          required=""
          type="text"
          value=""
        />
      </div>
      <input
        class="btn btn-primary btn-sm"
        data-testid="submit"
        type="submit"
        value="Add"
      />
    </form>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            na-na
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Store 4`] = `
<DocumentFragment>
  <div
    class="col-5"
  >
    <form
      action=""
      class="form-inline"
    >
      <div
        class="form-group mx-sm-3"
      >
        <input
          data-testid="input"
          required=""
          type="text"
          value="another task"
        />
      </div>
      <input
        class="btn btn-primary btn-sm"
        data-testid="submit"
        type="submit"
        value="Add"
      />
    </form>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            na-na
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Store 5`] = `
<DocumentFragment>
  <div
    class="col-5"
  >
    <form
      action=""
      class="form-inline"
    >
      <div
        class="form-group mx-sm-3"
      >
        <input
          data-testid="input"
          required=""
          type="text"
          value=""
        />
      </div>
      <input
        class="btn btn-primary btn-sm"
        data-testid="submit"
        type="submit"
        value="Add"
      />
    </form>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            another task
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            na-na
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Store 6`] = `
<DocumentFragment>
  <div
    class="col-5"
  >
    <form
      action=""
      class="form-inline"
    >
      <div
        class="form-group mx-sm-3"
      >
        <input
          data-testid="input"
          required=""
          type="text"
          value=""
        />
      </div>
      <input
        class="btn btn-primary btn-sm"
        data-testid="submit"
        type="submit"
        value="Add"
      />
    </form>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            na-na
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
